File: packages/shared/src/inbox.ts
import { z } from "zod";
import { itemKindSchema } from "./item";

export const capturePipelineStatusSchema = z.enum([
  "captured",
  "queued",
  "processing",
  "ready",
  "failed",
  "applied",
  "archived",
]);

export const captureTypeSchema = z.enum([
  "text",
  "voice",
  "photo",
  "link",
  "file",
  "share",
  "deeplink",
]);

export const captureStatusSchema = z.enum([
  "draft",
  "captured",
  "queued",
  "processing",
  "ready",
  "failed",
  "applied",
  "archived",
]);

export const captureActionTypeSchema = z.enum([
  "create_event",
  "create_task",
  "create_item",
  "draft_reply",
  "pay_invoice",
  "summarize",
  "review",
]);

export const captureCategorySchema = z.enum([
  "finance",
  "communication",
  "schedule",
  "document",
  "travel",
  "personal",
  "general",
]);

export const captureActorSchema = z.enum(["user", "sync", "system"]);

export const captureBadgeKindSchema = z.enum([
  "type",
  "category",
  "actor",
  "tag",
  "status",
]);

export const captureBadgeSchema = z.object({
  key: z.string().min(1),
  label: z.string().min(1),
  kind: captureBadgeKindSchema,
  tone: z.enum(["default", "secondary", "outline"]).default("outline"),
});

const metadataRecordSchema = z.record(z.string(), z.unknown());

export const captureActionSuggestionSchema = z.object({
  key: z.string().min(1),
  label: z.string().min(1),
  type: captureActionTypeSchema,
  confidence: z.number().min(0).max(1),
  requires_confirm: z.boolean(),
  preview_payload: metadataRecordSchema.default({}),
  is_primary: z.boolean().default(false),
});

const inboxCaptureOutWireSchema = z.object({
  id: z.string().uuid(),
  item_id: z.string().uuid().nullable().optional(),
  raw_content: z.string(),
  source: z.string().nullable().optional(),
  source_type: z.string().optional(),
  capture_type: captureTypeSchema,
  status: captureStatusSchema,
  pipeline_state: capturePipelineStatusSchema.optional(),
  treated_bucket: z.enum(["untreated", "treated"]).optional(),
  agent_hint: z.string().nullable().optional(),
  coming_soon_capabilities: z.array(z.string()).optional(),
  metadata: metadataRecordSchema.optional(),
  meta: metadataRecordSchema.optional(),
  suggested_actions: z.array(captureActionSuggestionSchema).default([]),
  primary_action: captureActionSuggestionSchema.nullable().optional(),
  requires_review: z.boolean().optional(),
  archived: z.boolean().optional(),
  archived_reason: z.enum(["applied", "deleted", "archived"]).nullable().optional(),
  deleted_at: z.string().datetime().nullable().optional(),
  category: captureCategorySchema.nullable().optional(),
  actor: captureActorSchema.optional(),
  tags: z.array(z.string()).optional(),
  badges: z.array(captureBadgeSchema).optional(),
  created_at: z.string().datetime(),
});

export const inboxCaptureOutSchema = inboxCaptureOutWireSchema.transform(
  ({
    meta,
    metadata,
    requires_review,
    primary_action,
    suggested_actions,
    item_id,
    source_type,
    pipeline_state,
    treated_bucket,
    agent_hint,
    coming_soon_capabilities,
    archived,
    archived_reason,
    deleted_at,
    category,
    actor,
    tags,
    badges,
    ...rest
  }) => ({
    ...rest,
    item_id: item_id ?? null,
    source_type: source_type ?? rest.capture_type,
    pipeline_state: pipeline_state ?? rest.status,
    treated_bucket: treated_bucket ?? (rest.status === "applied" ? "treated" : "untreated"),
    agent_hint: agent_hint ?? null,
    coming_soon_capabilities: coming_soon_capabilities ?? [],
    metadata: metadata ?? meta ?? {},
    suggested_actions,
    primary_action: primary_action ?? null,
    requires_review: requires_review ?? false,
    archived: archived ?? false,
    archived_reason: archived_reason ?? null,
    deleted_at: deleted_at ?? null,
    category: category ?? null,
    actor: actor ?? "user",
    tags: tags ?? [],
    badges: badges ?? [],
  })
);

export const inboxListResponseSchema = z
  .object({
    captures: z.array(inboxCaptureOutSchema).optional(),
    entries: z.array(inboxCaptureOutSchema).optional(),
  })
  .transform(({ captures, entries }) => {
    const resolvedEntries = entries ?? captures ?? [];
    const resolvedCaptures = captures ?? entries ?? [];
    return {
      captures: resolvedCaptures,
      entries: resolvedEntries,
    };
  });

export const createCaptureRequestSchema = z.object({
  raw_content: z.string().default(""),
  source: z.string().nullable().optional().default("manual"),
  capture_type: captureTypeSchema.default("text"),
  status: captureStatusSchema.default("captured"),
  metadata: metadataRecordSchema.default({}),
});

export const captureUploadResponseSchema = z.object({
  id: z.string().uuid(),
  status: captureStatusSchema,
});

const captureAssetWireSchema = z.object({
  id: z.string().uuid(),
  kind: z.string(),
  storage_key: z.string(),
  mime_type: z.string(),
  size_bytes: z.number().int().nonnegative(),
  duration_ms: z.number().int().nullable().optional(),
  checksum: z.string().nullable().optional(),
  metadata: metadataRecordSchema.optional(),
  meta: metadataRecordSchema.optional(),
  created_at: z.string().datetime(),
  preview_kind: z.enum(["audio", "image", "pdf", "text", "binary"]).optional(),
  file_name: z.string().optional(),
  content_path: z.string().optional(),
});

export const captureAssetOutSchema = captureAssetWireSchema.transform(
  ({ meta, metadata, preview_kind, file_name, content_path, ...rest }) => ({
    ...rest,
    metadata: metadata ?? meta ?? {},
    duration_ms: rest.duration_ms ?? null,
    checksum: rest.checksum ?? null,
    preview_kind: preview_kind ?? "binary",
    file_name: file_name ?? "",
    content_path: content_path ?? "",
  })
);

export const captureArtifactOutSchema = z.object({
  id: z.string().uuid(),
  artifact_type: z.string(),
  content_json: metadataRecordSchema,
  provider: z.string().nullable().optional(),
  model: z.string().nullable().optional(),
  confidence: z.number().nullable().optional(),
  created_at: z.string().datetime(),
});

export const captureJobOutSchema = z.object({
  id: z.string().uuid(),
  job_type: z.string(),
  status: z.string(),
  attempt_count: z.number().int().nonnegative(),
  last_error: z.string().nullable().optional(),
  scheduled_at: z.string().datetime().nullable().optional(),
  started_at: z.string().datetime().nullable().optional(),
  finished_at: z.string().datetime().nullable().optional(),
  created_at: z.string().datetime(),
});

export const captureDetailResponseSchema = z.object({
  capture: inboxCaptureOutSchema,
  assets: z.array(captureAssetOutSchema),
  artifacts: z.array(captureArtifactOutSchema),
  jobs: z.array(captureJobOutSchema),
  pipeline_trace: z.array(metadataRecordSchema).default([]),
  artifacts_summary: metadataRecordSchema.default({}),
});

export const captureArtifactsResponseSchema = z.object({
  artifacts: z.array(captureArtifactOutSchema),
});

export const reprocessCaptureResponseSchema = z.object({
  capture_id: z.string().uuid(),
  status: z.string(),
});

export const processCaptureRequestSchema = z.object({
  title: z.string().min(1),
});

export const processCaptureResponseSchema = z.object({
  item_id: z.string().uuid(),
});

export const capturePreviewRequestSchema = z.object({
  action_key: z.string().optional(),
});

export const capturePreviewResponseSchema = z.object({
  capture_id: z.string().uuid(),
  action_key: z.string().nullable().optional(),
  action_type: captureActionTypeSchema.nullable().optional(),
  suggested_title: z.string(),
  suggested_kind: itemKindSchema,
  confidence: z.number(),
  reason: z.string(),
  preview_payload: metadataRecordSchema.default({}),
});

export const applyCaptureRequestSchema = z.object({
  title: z.string().optional(),
  kind: itemKindSchema.optional(),
  action_key: z.string().optional(),
  metadata: metadataRecordSchema.default({}),
});

export const applyCaptureResponseSchema = z.object({
  capture_id: z.string().uuid(),
  item_id: z.string().uuid(),
  event_id: z.string().uuid().nullable().optional(),
  applied_action_key: z.string().nullable().optional(),
});

export const captureActionResponseSchema = z.object({
  capture_id: z.string().uuid(),
  status: z.string(),
});

export const externalCaptureRequestSchema = z.object({
  raw_content: z.string().default(""),
  source: z.string().nullable().optional().default("extension"),
  capture_type: captureTypeSchema.default("text"),
  metadata: metadataRecordSchema.default({}),
  idempotency_key: z.string().nullable().optional(),
});

export type CapturePipelineStatus = z.infer<typeof capturePipelineStatusSchema>;
export type CaptureType = z.infer<typeof captureTypeSchema>;
export type CaptureStatus = z.infer<typeof captureStatusSchema>;
export type CaptureActionType = z.infer<typeof captureActionTypeSchema>;
export type CaptureCategory = z.infer<typeof captureCategorySchema>;
export type CaptureActor = z.infer<typeof captureActorSchema>;
export type CaptureBadge = z.infer<typeof captureBadgeSchema>;
export type CaptureActionSuggestion = z.infer<typeof captureActionSuggestionSchema>;
export type InboxCaptureOut = z.infer<typeof inboxCaptureOutSchema>;
export type InboxListResponse = z.infer<typeof inboxListResponseSchema>;
export type CreateCaptureRequest = z.input<typeof createCaptureRequestSchema>;
export type CaptureUploadResponse = z.infer<typeof captureUploadResponseSchema>;
export type CaptureAssetOut = z.infer<typeof captureAssetOutSchema>;
export type CaptureArtifactOut = z.infer<typeof captureArtifactOutSchema>;
export type CaptureJobOut = z.infer<typeof captureJobOutSchema>;
export type CaptureDetailResponse = z.infer<typeof captureDetailResponseSchema>;
export type CaptureArtifactsResponse = z.infer<typeof captureArtifactsResponseSchema>;
export type ReprocessCaptureResponse = z.infer<typeof reprocessCaptureResponseSchema>;
export type ProcessCaptureRequest = z.input<typeof processCaptureRequestSchema>;
export type ProcessCaptureResponse = z.infer<typeof processCaptureResponseSchema>;
export type CapturePreviewRequest = z.input<typeof capturePreviewRequestSchema>;
export type CapturePreviewResponse = z.infer<typeof capturePreviewResponseSchema>;
export type ApplyCaptureRequest = z.input<typeof applyCaptureRequestSchema>;
export type ApplyCaptureResponse = z.infer<typeof applyCaptureResponseSchema>;
export type CaptureActionResponse = z.infer<typeof captureActionResponseSchema>;
export type ExternalCaptureRequest = z.input<typeof externalCaptureRequestSchema>;

File: apps/api/src/schemas/inbox.py
import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field
from src.schemas.item import EntityLinkOut

CaptureType = Literal["text", "voice", "photo", "link", "file", "share", "deeplink"]
CaptureStatus = Literal[
    "draft",
    "captured",
    "queued",
    "processing",
    "ready",
    "failed",
    "applied",
    "archived",
]
CaptureActionType = Literal[
    "create_event",
    "create_task",
    "create_item",
    "draft_reply",
    "pay_invoice",
    "review",
]
CaptureCategory = Literal[
    "finance",
    "communication",
    "schedule",
    "document",
    "travel",
    "personal",
    "general",
]
CaptureActor = Literal["user", "sync", "system"]
CaptureBadgeKind = Literal["type", "category", "actor", "tag", "status"]
ItemKind = Literal["note", "objective", "task", "resource"]
TreatedBucket = Literal["untreated", "treated"]


class CaptureActionSuggestionOut(BaseModel):
    key: str
    label: str
    type: CaptureActionType
    confidence: float
    requires_confirm: bool
    preview_payload: dict[str, Any] = Field(default_factory=dict)
    is_primary: bool = False


class CaptureBadgeOut(BaseModel):
    key: str
    label: str
    kind: CaptureBadgeKind
    tone: Literal["default", "secondary", "outline"] = "outline"


class InboxCaptureOut(BaseModel):
    id: uuid.UUID
    item_id: uuid.UUID | None = None
    title: str | None = None
    raw_content: str
    source: str | None = None
    source_type: str = "text"
    capture_type: CaptureType
    status: CaptureStatus
    pipeline_state: CaptureStatus = "captured"
    treated_bucket: TreatedBucket = "untreated"
    agent_hint: str | None = None
    coming_soon_capabilities: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(
        alias="meta",
        serialization_alias="metadata",
    )
    suggested_actions: list[CaptureActionSuggestionOut] = Field(default_factory=list)
    primary_action: CaptureActionSuggestionOut | None = None
    requires_review: bool = False
    archived: bool = False
    archived_reason: Literal["applied", "deleted", "archived"] | None = None
    deleted_at: datetime | None = None
    category: CaptureCategory | None = None
    actor: CaptureActor = "user"
    tags: list[str] = Field(default_factory=list)
    badges: list[CaptureBadgeOut] = Field(default_factory=list)
    created_at: datetime

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class InboxListResponse(BaseModel):
    captures: list[InboxCaptureOut]
    entries: list[InboxCaptureOut] = Field(default_factory=list)


class InboxSearchEntryOut(BaseModel):
    id: uuid.UUID
    title: str
    capture_type: CaptureType
    status: CaptureStatus
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class InboxSearchResponse(BaseModel):
    captures: list[InboxSearchEntryOut]


class CaptureLinksResponse(BaseModel):
    links: list[EntityLinkOut]


class CreateCaptureRequest(BaseModel):
    raw_content: str = ""
    source: str | None = "manual"
    capture_type: CaptureType = "text"
    status: CaptureStatus = "captured"
    metadata: dict[str, Any] = Field(default_factory=dict)


class UpdateCaptureRequest(BaseModel):
    title: str | None = None


class CaptureUploadResponse(BaseModel):
    id: uuid.UUID
    status: CaptureStatus
    task_id: str | None = None
    run_id: uuid.UUID | None = None
    queue_state: Literal["enqueued", "not_enqueued"] | None = None
    queue_name: str | None = None


class CaptureAssetOut(BaseModel):
    id: uuid.UUID
    kind: str
    storage_key: str
    mime_type: str
    size_bytes: int
    duration_ms: int | None = None
    checksum: str | None = None
    metadata: dict[str, Any] = Field(
        alias="meta",
        serialization_alias="metadata",
    )
    created_at: datetime
    preview_kind: Literal["audio", "image", "pdf", "text", "binary"] = "binary"
    file_name: str = ""
    content_path: str = ""

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class CaptureArtifactOut(BaseModel):
    id: uuid.UUID
    artifact_type: str
    content_json: dict[str, Any]
    provider: str | None = None
    model: str | None = None
    confidence: float | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class CaptureJobOut(BaseModel):
    id: uuid.UUID
    job_type: str
    run_id: uuid.UUID | None = None
    queue_name: str | None = None
    task_id: str | None = None
    status: str
    attempt_count: int
    last_error: str | None = None
    scheduled_at: datetime | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class CaptureDetailResponse(BaseModel):
    capture: InboxCaptureOut
    assets: list[CaptureAssetOut]
    artifacts: list[CaptureArtifactOut]
    jobs: list[CaptureJobOut]
    pipeline_trace: list[dict[str, Any]] = Field(default_factory=list)
    artifacts_summary: dict[str, Any] = Field(default_factory=dict)


class CaptureArtifactsResponse(BaseModel):
    artifacts: list[CaptureArtifactOut]


class ProcessCaptureRequest(BaseModel):
    title: str


class ProcessCaptureResponse(BaseModel):
    item_id: uuid.UUID


class CapturePreviewRequest(BaseModel):
    action_key: str | None = None


class CapturePreviewResponse(BaseModel):
    capture_id: uuid.UUID
    action_key: str | None = None
    action_type: CaptureActionType | None = None
    suggested_title: str
    suggested_kind: ItemKind
    confidence: float
    reason: str
    preview_payload: dict[str, Any] = Field(default_factory=dict)
    missing_fields: list[str] = Field(default_factory=list)


class ApplyCaptureRequest(BaseModel):
    title: str | None = None
    kind: ItemKind | None = None
    action_key: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class ApplyCaptureResponse(BaseModel):
    capture_id: uuid.UUID
    item_id: uuid.UUID
    event_id: uuid.UUID | None = None
    applied_action_key: str | None = None


class CaptureActionResponse(BaseModel):
    capture_id: uuid.UUID
    status: str
    task_id: str | None = None
    run_id: uuid.UUID | None = None
    queue_state: Literal["enqueued", "not_enqueued"] | None = None
    queue_name: str | None = None


class NoteSummaryRefreshResponse(BaseModel):
    capture_id: uuid.UUID
    status: Literal["generated", "unchanged", "skipped_too_short"]
    summary_updated: bool = False
    source_hash: str | None = None


class ExternalCaptureRequest(BaseModel):
    raw_content: str = ""
    source: str | None = "extension"
    capture_type: CaptureType = "text"
    metadata: dict[str, Any] = Field(default_factory=dict)
    idempotency_key: str | None = None
